### PR TITLE
[VERSION_APITEST] Add tests for VerQueryValue(A/W)

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Add tests for VerQueryValue(A/W)

JIRA issue: [CORE-19783](https://jira.reactos.org/browse/CORE-19783)

## Tests
- [ ] Test WHS: tbd
- [ ] Test Windows 2003 x64: tbd

### CODE REVIEW

**Comment by whindsaks** [comment]
Does this hardcoded Test_StaticVersionnfo reproduce the problem seen in the Jira issue? Meaning, do some of the VerQueryValue tests here fail with current ROS version.dll?

**Comment by Doug-Lyons** [comment]
> Does this hardcoded Test_StaticVersionnfo reproduce the problem seen in the Jira issue? Meaning, do some of the VerQueryValue tests here fail with current ROS version.dll?

@whindsaks, Yes, it has four (4) failures on ReactOS current master (and wine 9.3). These are fixed by the most recent patches in https://jira.reactos.org/browse/CORE-19783.

**Review by Doug-Lyons — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]

File: modules/rostests/apitests/version/testlist.c [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#define __ROS_LONG64__
```
Comment (binarymaster) [inline modules/rostests/apitests/version/testlist.c:1]:
```suggestion
/*
 * PROJECT:     ReactOS API Tests
 * LICENSE:     LGPL-2.0-or-later (https://spdx.org/licenses/LGPL-2.0-or-later)
 * PURPOSE:     Tests for version.dll
 * COPYRIGHT:   Copyright 2024 Timo Kreuzer <timo.kreuzer@reactos.org>
 */
 
#define __ROS_LONG64__
```

**Review by HBelusca — APPROVED** [review]
Minor nitpicks and some questions, but ok otherwise. 
I'm curious about the x64 results.

File: modules/rostests/apitests/version/VerQueryValue.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+static const TEST_VERSIONINFO g_VersionInfo =
+{
+    .wLength = 298, // Header: 40, Value: 54, StringFileInfo: 204
+    .wValueLength = sizeof(VS_FIXEDFILEINFO),
+    .wType = 0,
+    .szKey = L"VS_VERSION_INFO",
+    .Padding1 = 0,
+    .Value =
+    {
+        .dwSignature = 0xFEEF04BD,
```
[the remark this answers is not in the document]
Comment (whindsaks) [reply modules/rostests/apitests/version/VerQueryValue.c:82]: Yes, [VS_FIXEDFILEINFO::dwSignature](https://learn.microsoft.com/en-us/windows/win32/api/verrsrc/ns-verrsrc-vs_fixedfileinfo)

Comment (HBelusca) [inline modules/rostests/apitests/version/VerQueryValue.c:82]:
```suggestion
        .dwSignature = VS_FFI_SIGNATURE,
```

File: modules/rostests/apitests/version/VerQueryValue.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        .dwStrucVersion = 0x00010000,
```
Comment (HBelusca) [inline modules/rostests/apitests/version/VerQueryValue.c:83]:
```suggestion
        .dwStrucVersion = VS_FFI_STRUCVERSION,
```

File: modules/rostests/apitests/version/VerQueryValue.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        .dwFileVersionMS = 0x00050002,
+        .dwFileVersionLS = 0x0ECE0F77,
+        .dwProductVersionMS = 0x00040009,
+        .dwProductVersionLS = 0x00040004,
+        .dwFileFlagsMask = 0x00000002,
+        .dwFileFlags = 0x00000000,
+        .dwFileOS = 0x00000004,
+        .dwFileType = 0x00000001,
```
Comment (HBelusca) [inline modules/rostests/apitests/version/VerQueryValue.c:91]:
```suggestion
        .dwFileType = VFT_APP,
```

File: modules/rostests/apitests/version/VerQueryValue.c [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    {
+        .dwSignature = 0xFEEF04BD,
+        .dwStrucVersion = 0x00010000,
+        .dwFileVersionMS = 0x00050002,
+        .dwFileVersionLS = 0x0ECE0F77,
+        .dwProductVersionMS = 0x00040009,
+        .dwProductVersionLS = 0x00040004,
+        .dwFileFlagsMask = 0x00000002,
+        .dwFileFlags = 0x00000000,
+        .dwFileOS = 0x00000004,
```
Comment (HBelusca) [inline modules/rostests/apitests/version/VerQueryValue.c:90]: Technically this is `VOS__WINDOWS32`, but I'm winding whether you could use `VOS_NT_WINDOWS32` there (equals to 0x00040004)

File: modules/rostests/apitests/version/VerQueryValue.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    BOOL result;
+
+    /* Test a special static version */
+    result = VerQueryValueW(&g_VersionInfo,
+                            L"\\StringFileInfo\\000004B0\\CompanyName",
+                            &pvData,
+                            &cbLen);
+    ok(result, "VerQueryValueW failed\n");
+    ok_eq_int(cbLen, 16);
+    ok(wcscmp((WCHAR*)pvData, L"ReactOS Project") == 0, "Unexpected value\n");
```
Comment (learn-more) [inline modules/rostests/apitests/version/VerQueryValue.c:252]: in `apitest.h` we defined `ok_eq_wstr` that might be of use here:
`#define ok_eq_wstr(value, expected)         ok(!wcscmp(value, expected), #value " = \"%ls\", expected \"%ls\"\n", value, expected)`

**Review by learn-more — APPROVED** [review]

File: modules/rostests/apitests/version/VerQueryValue.c [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    result = VerQueryValueW(&g_VersionInfo,
+                            L"\\StringFileInfo\\000004B0\\FooBar",
+                            &pvData,
+                            &cbLen);
+    ok(result, "VerQueryValueW failed\n");
+    ok_eq_int(cbLen, 4);
+    ok(wcscmp((WCHAR*)pvData, L"Bar") == 0, "Bar\n");
+    ExpectedOffset = FIELD_OFFSET(TEST_VERSIONINFO, StringFileInfo.StringTable.FooBar.Value);
+    ok(pvData == (PVOID)((ULONG_PTR)&g_VersionInfo + ExpectedOffset), "Unexpected offset\n");
+
```
Comment (JoachimHenze) [inline modules/rostests/apitests/version/VerQueryValue.c:285]:
```suggestion
```

File: modules/rostests/apitests/version/testlist.c [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+#define STANDALONE
+#include <apitest.h>
+
+extern void func_VerQueryValue(void);
+
+const struct test winetest_testlist[] =
+{
+    { "VerQueryValue",                 func_VerQueryValue },
+
```
Comment (JoachimHenze) [inline modules/rostests/apitests/version/testlist.c:11]: Does this empty line serve any purpose that I missed?
```suggestion
```

**Comment by tkreuzer** [comment]
GitHub fubared the branch name, closing and reopening.

**Comment by SergeGautherie** [comment]
> reopening

as #7473.

**Pull request closed without merging**